Register types related to K8s Ingress

michi-covalent opened the pull request:
please review commit by commit:

- 689581540b1de6d03d089bdfbd61e9c18d4f1aa1 Add types related to K8s Ingress.
- 7186ac70e4e0efeb7d989fe806564246e4456777 Generate k8s client and register ingress types.

jrajahalme committed:
k8s: Add CiliumEnvoyConfig CRD

Add a new CiliumEnvoyConfig to cilium.io API v2alpha1

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Switch to protobuf Go APIv2

Switch to protobuf Go APIv2 by changing imports from
'github.com/golang/protobuf' to 'google.golang.org/protobuf'.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Call 'callback' also when using current resource version

Call the completion callback also when there is no change in the
cached resource and current version is used, but it has not been
acknowledged yet.

In current use the prior behavior did not cause a problem, but the new
Envoy CRD code depends on the callback to be called also in this case.

Note that the callback is NOT called when there is
no change in the resource and the current version has already been
acknowledged.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Add xDS support for RDS, CDS, & EDS

Add support for Route, Cluster, and Endpoint discovery services.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
lb: Add support for L7 load balancers

Add new SVC_FLAG_L7LOADBALANCER to designate services that are not
loadbalanced in BPF, but are forwarded to a local TPROXY loadbalancer
instead.

The set of proxy ports registered to datapath was limited to the
preconfigured set of HTTP, DNS, and proxylib ports. Allow adding (CRD)
proxy ports to this set.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
datapath: Pass proxy port in to-proxy traces

Include proxy port in the to-proxy trace messages.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
service: Push Envoy Endpoints for L7 LB k8s services

Upsert Envoy endpoint resources for services that have been selected
for L7 LB in CiliumEnvoyConfig CRDs.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Add xDS resource validation

Make sure Envoy xDS resources have names and validate xDS resources
before adding them to xDS cache. This validation can not be done at
the k8s CRD apply time, so k8s CRD can contain invalid Envoy
configurations.

Envoy Go API is updated to contain the generated validation code.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Allocate CRD listener ports if unspecified

Add support for dynamically created (CRD) proxy ports in pkg/proxy.

Allocate proxy ports for CRD listeners without Listener.Address
specification and install datapath TPROXY rules if
Listener.Address is left empty.

Allocated proxy ports are sticky in order to return the same port
when the same listener (by name) is reconfigured or configured again
after deletion.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Register k8s service redirection to CRD listener

Register L7 LB service redirections. Use 'L7LBInfo' to keep track of
the CiliumEnvoyConfig that requested a service to be registered for L7
load balancing. Only one CiliumEnvoyConfig may request redirection of
a given service at a time, as otherwise the choice between Envoy
listeners would be ambiguous.

Service backends are synced to Envoy, so that the ingress service can
load balance to them. Datapath L7LB redirection is only requested for
'services' in a CiliumEnvoyResources. Backends are synced to Envoy
also for 'backendServices'. This allows the regular eBPF LB to be used
for pod-to-pod traffic for 'backendServices' while L7 LB is used for
ingress to those same services.

Note: 'Ingress' CEC configurations are still experimental. 

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
bpf: Add support for re-entering LXC egress path after L7 LB

Allow host to tail call to lxc egress for policy enforcement after L7 LB.

This requires L7 LB proxy to set the source Endpoint ID in the socket
mark. This mark includes a new magic value that the datapath
recognizes to subject the traffic to source LXC (policy enforcement)
processing again. Note that this does not loop endlessly due to the L7
LB switching from service frontend IP to backend IP.

Envoy image is updated to support the new EndpointId option for the
bpf_metadata listener filter.

iptables rules are amended to avoid overwriting the new "from L7 LB"
mark with "from host" mark.

A new flag 'from_l7lb' is added to conntrack to enable redirecting
reply packets back to L7 LB proxy.

Add a new map for bpf_host to tail call into for LXC policy
enforcement after L7 LB.

Some kernels reached their eBPF complexity limit with the new code
compiled in. Address this by making all L7 LB related code in
bpf_lxc.c confitional on the new ENABLE_L7_LB. This allows the new
feature to be used on kernels that allow a but more complex eBPF
programs.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Update to add SO_REUSEADDR option to upstream connections.

Envoy retries may fail on "address already in use" when the original
source address and port are used on upstream connections. Cilium
typically does this in the egress proxy listeners. Fix this by using a
Cilium Envoy build that always sets SO_REUSEADDR when original source
address and port is used.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
iptables: Add notrack for L7 LB upstream connections

L7 load balancer upstream connections may go to the egress proxy
listener of the source pod. Pod connections to proxy listeners are
already not tracked to avoid conflicts due to reuse of the same
5-tuple on both sides of the (egress) proxy. Do the same for L7 LB
connections that may go to the egress proxy for the same reason.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Clarify xDS revert function locking

xDS revert functions can not be called from the callback functions
provided to xDS cache update functions due to a locking that would
lead to a deadlock. Revert functions are designed to be called after a
larger transaction fails, not after a single resource update
fails. This is why the revert functions assume to be called outside of
any given callback function and therefore take the locks needed for
that.

Clarify this and add a type that calls the revert functions n reverse
order.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

michi-covalent committed:
Add types related to K8s Ingress

This commit adds K8s types related to Ingress. The current plan is to
implement an ingress controller logic in cilium-operator that translates
ingress definitions to corresponding CiliumEnvoyConfig custom resources.

Ref: #17191

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

michi-covalent committed:
Generate k8s client and register Ingress types

Re-generate k8s client by running `make generate-k8s-api` and register
Ingress and IngressClass types.

Ref: #17191

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

aanm commented on pkg/k8s/slim/k8s/api/networking/v1/types.go:
```diff
@@ -191,3 +191,349 @@ type NetworkPolicyList struct {
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// IngressClassList is a collection of IngressClasses.
+type IngressClassList struct {
```
This, similarly to `IngressClass`, `Ingress` and `IngressList`, needs to be registered in the file pkg/k8s/slim/k8s/api/networking/v1/register.go

michi-covalent replied:
yeah i did that in the next commit: https://github.com/cilium/cilium/blob/7186ac70e4e0efeb7d989fe806564246e4456777/pkg/k8s/slim/k8s/api/networking/v1/register.go#L35

would it be easier to review if it's a single commit?

aanm reviewed (changes requested):
Just double checking, will we use all fields of all structures? If no, might make sense to remove certain fields.

Also, it would be nice to have the code that will use these structures in this same PR instead of different PRs.

jrajahalme committed:
envoy: Do not wait for RDS, CDS, EDS, or SDS

Do not wait for routes (RDS), clusters (CDS), cluster load assignments
(EDS), or secrets (SDS) as they will only be acked if referenced from
a listener that may not exist yet or may have been deleted already.

Add Cilium agent option --envoy-config-timeout to limit the maximum
time the agent waits for Envoy (Listener) configuration to be acked,
defaulting to 2 minutes. This unblocks k8s watchers if the agent ever
waits for Envoy for resources that are not being acknoowledged.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

michi-covalent committed:
add service

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

jrajahalme committed:
datapath: Redirect nodeport L7 LB services

Redirect L7 LB nodeport service connections received by the host to an
Envoy listener. Prior to this commit these services worked as intended
only when the service connection was originated by a pod. This commit
implements L7 LB redirection for nodeport services received by the
node from other nodes or from an external load balancer.

Redirect to nodeport L7 LB from bpf_sock for hostns sources. This is
needed due to this case not being handled by eBPF datapath on TC
level. Note that this requires a recent kernel to work (Linux
5.12+). With older kernels nodeport services only work between nodes,
not from a node host namespace to the node itself.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Initial SDS support

Add Initial support for Secret Discovery Service. Cilium xds cluster
details need to be given on each instance, omitting SDS config source
is not yet supported.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
datapath: Add ctx_redirect_to_proxy_hairpin_ipv6

Add IPv6 version of ctx_redirect_to_proxy_hairpin() and use it in IPv6 path.

Fixes: #17717
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

michi-covalent committed:
add DeletionTimestamp

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

michi-covalent committed:
generate client

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

jrajahalme commented on operator/watchers/envoy_config.go:
```diff
@@ -0,0 +1,216 @@
+			},
+			BackendServices: backendServices,
+			// TODO(michi) what is this
+			// Ingress:         true,
```
This is a flag for *proxy ingress to pod* vs. *proxy egress from pod*. This is not to be used with k8s Ingress :-). Also, it is yet to be seen if we will ever actually need CiliumEnvoyConfig for pod ingress. So far it is completely untested and the bpf datapath support for it is also missing.
